### ISSUE TITLE
Generic with `ParamSpec` cannot be assigned to same generic with ellipsis

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#2730]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

from typing import Generic, Protocol, TypeVar
from typing_extensions import ParamSpec

P = ParamSpec('P')
T = TypeVar('T')


class Handler(Protocol[P]):
    async def __call__(self, /, *args: P.args, **kwargs: P.kwargs) -> None:
        ...


class Command(Generic[P]):
    def __init__(self, handler: Handler[P]) -> None:
        self.handler: Handler[P] = handler


commands: list[Command[...]] = []


def do_something(int_handler: Handler[int], var_args_handler: Handler[P], /) -> None:
    int_command = Command(int_handler)
    commands.append(int_command)

    var_args_command = Command(var_args_handler)
    commands.append(var_args_command)
```

An error is generated at line 28:

```
Argument of type "Command[P@do_something]" cannot be assigned to parameter "__object" of type "Command[(...)]" in function "append"
    TypeVar "P@Command" is invariant
      Type "P@do_something" cannot be assigned to type "(...)"
```

**Expected behavior**
No error should be generated.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line

Which version?
1.1.197

**Comment by erictraut** [comment]
I think you might be using ParamSpecs in cases that are beyond their intended use.

I tried this code in pyre, which is the reference implementation for ParamSpec (since the authors of PEP 612 work on pyre). It generates the same error as pyright in this circumstance, which is evidence that the authors either didn't consider this case when they were writing the PEP or don't consider this case as valid.

I could make the argument either for or against supporting this. The argument for it is based on consistency with the way `Callable[..., x]` works. The argument against it is that the use case doesn't make much sense. If you're concerned enough about type checking that you're using ParamSpec, then it seems odd that you would use a `...` signature because it throws out all type checking. If your intent is to support "any signature", then why not use `Callable[..., x]`?

**Comment by bryanforbes** [comment]
The code above is a really arbitrary example, so there are probably ways around the error for the given code. The issue cropped up in the library I'm developing where the first argument of all handlers is a context object that needs to be a specific type for certain command types [Here's a more complex example that's closer to what my library is doing](https://gist.github.com/bryanforbes/201ef374aebdbe7179bedf077af85021). However after looking at this new example in the gist, I see a couple of ways I can get around the error I'm encountering:

* Use a protocol instead of `Command` in the type for `Application.commands`
* Use `...` instead of `P` in the declaration for `Thang`

It's quite possible I'm using `ParamSpec` in a way that wasn't considered when the PEP was written or is invalid, so I'm willing to admit what I'm attempting to do is wrong; but my gut feeling is my original approach should still be valid (of course, my gut has been wrong before). Is there anyone we can pull into this conversation to get their take on it?

**Comment by erictraut** [comment]
We could pull in @mrkmndz, @shannonzhu, @pradeep90, or @stroxler. They work on pyre. Mark was the primary author of PEP 612. I don't know if they'll receive these notifications though. You could try filing an issue in the [pyre github repo](https://github.com/facebook/pyre-check).

**Comment by erictraut** [comment]
I've implemented the logic to support this. I'm still not 100% convinced that this should be allowed, but there's a reasonable argument that it should.

**Comment by erictraut** [comment]
This is included in pyright 1.1.198, which I just published. It will also be included in the next release of pylance.

**Comment by pradeep90** [comment]
@erictraut This is a bug in Pyre where `Foo[...]` is parsed incorrectly. Apart from that, the idea works as expected. There is conceptually no blocker to using `...` like: `x: Command[int]; y: Command[...] = x`
